controller/apps.go

```diff
@@ -0,0 +1,443 @@
+	// use WIT API to convert Space UUID to Space name
+	oc := c.getAndCheckOSIOClient(ctx)
+
+	osioSpace, err := oc.GetSpaceByID(ctx, spaceID)
+	if err != nil {
+		return nil, err
+	}
+	return osioSpace.Attributes.Name, nil
+}
+
+func (c *AppsController) getNamespaceName(ctx context.Context) (*string, error) {
+
+	osioclient := c.getAndCheckOSIOClient(ctx)
+	kubeSpaceAttr, err := osioclient.GetNamespaceByType(ctx, nil, "user")
+	if err != nil {
+		return nil, err
+	}
+	if kubeSpaceAttr == nil || kubeSpaceAttr.Name == nil {
+		return nil, witerrors.NewNotFoundError("namespace", "user")
+	}
+
+	return kubeSpaceAttr.Name, nil
+}
+
+func (c *AppsController) getUser(authClient authservice.Client, ctx context.Context) (*authservice.User, error) {
+	// get the user definition (for cluster URL)
+	resp, err := authClient.ShowUser(ctx, authservice.ShowUserPath(), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	var respType authservice.User
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+func (c *AppsController) getToken(authClient authservice.Client, ctx context.Context, forService string) (*authservice.TokenData, error) {
+
+	resp, err := authClient.RetrieveToken(ctx, authservice.RetrieveTokenPath(), forService, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	var respType authservice.TokenData
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+// getKubeClient createa kube client for the appropriate cluster assigned to the current user.
+// many different errors are possible, so controllers should call getAndCheckKubeClient() instead
+func (c *AppsController) getKubeClient(ctx context.Context) (*kubernetes.KubeClient, error) {
+
+	// create Auth API login object
+	authClient, err := auth.CreateClient(ctx, c.Config)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	authUser, err := c.getUser(*authClient, ctx)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	if authUser == nil || authUser.Data.Attributes.Cluster == nil {
+		log.Error(ctx, nil, "error getting user from Auth server:"+tostring(authUser))
+		return nil, errors.New("error getting user from Auth Server: %s" + tostring(authUser))
+	}
+
+	// get the login token for the cluster OpenShift API
+	osauth, err := c.getToken(*authClient, ctx, *authUser.Data.Attributes.Cluster)
```

Comment:
what kind of login token do you retrieve ? the arguments are a client, the context and a cluster.